src/leapfrogai_api/backend/rag/index.py

```
@@ -0,0 +1,113 @@
+embeddings_type: type[Embeddings] | type[LeapfrogAIEmbeddings] | None = (
+    LeapfrogAIEmbeddings
+)
+
+
+class FileAlreadyIndexedError(Exception):
+    pass
+
+
+class IndexingService:
+    """Service for indexing files into a vector store."""
+
+    def __init__(self, db: AsyncClient):
+        self.db = db
+        self.embeddings = embeddings_type()
+
+    async def index_file(self, vector_store_id: str, file_id: str) -> VectorStoreFile:
+        """Index a file into a vector store."""
+        crud_vector_store_file = CRUDVectorStoreFile(db=self.db)
+
+        if await crud_vector_store_file.get(
+            filters={"vector_store_id": vector_store_id, "id": file_id}
+        ):
+            raise FileAlreadyIndexedError("File already indexed")
+
+        crud_file_object = CRUDFileObject(db=self.db)
+        crud_file_bucket = CRUDFileBucket(db=self.db, model=UploadFile)
+
+        file_object = await crud_file_object.get(filters=FilterFileObject(id=file_id))
+
+        if not file_object:
+            raise ValueError("File not found")
+        file_bytes = await crud_file_bucket.download(id_=file_id)
+
+        with tempfile.NamedTemporaryFile(suffix=file_object.filename) as temp_file:
+            temp_file.write(file_bytes)
+            temp_file.seek(0)
+            documents = await load_file(temp_file.name)
+            chunks = await split(documents)
+
+            if len(chunks) == 0:
+                vector_store_file = VectorStoreFile(
+                    id=file_id,
+                    created_at=0,
+                    last_error=LastError(
+                        message="No text found in file", code="parsing_error"
+                    ),
+                    object="vector_store.file",
+                    status=VectorStoreFileStatus.FAILED.value,
+                    vector_store_id=vector_store_id,
+                )
+                return await crud_vector_store_file.create(object_=vector_store_file)
```

Is there a reason we want to store an empty file that failed to process?

It feels like we can raise an error here.